File: sales.js
const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');
const csv = require('csv-parser');
const nodeplotlib = require('nodeplotlib');

// Set seed for reproducibility
tf.setBackend('tensorflow');

// CustomScaler class
class CustomScaler {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }

    apply(data) {
        return tf.div(tf.sub(data, this.min), tf.sub(this.max, this.min));
    }

    inverseTransform(scaledData) {
        return tf.add(tf.mul(scaledData, tf.sub(this.max, this.min)), this.min);
    }

    // Method to get scalar data for saving
    getScalerData() {
        return { min: this.min.arraySync(), max: this.max.arraySync() };
    }
}

// Load and preprocess the dataset using csv-parser
async function loadDataset() {
    const rows = [];
    return new Promise((resolve, reject) => {
        fs.createReadStream('airline-passengers.csv')
            .pipe(csv())
            .on('data', (row) => rows.push(row))
            .on('end', () => {
                const values = rows.map(item => parseFloat(item.Passengers));
                const dataset = tf.tensor2d([values]).transpose();

                // Calculate min and max for custom scaler
                const min = tf.min(dataset);
                const max = tf.max(dataset);

                // Create custom scaler
                const scaler = new CustomScaler(min, max);

                // Apply custom scaler
                const normalizedData = scaler.apply(dataset);

                resolve({ dataset: normalizedData, scaler });
            })
            .on('error', (error) => reject(error));
    });
}

// Create the LSTM model
function createModel(inputShape) {
    const model = tf.sequential();

    model.add(tf.layers.lstm({ units: 16, inputShape }));
    model.add(tf.layers.dense({ units: 1 }));

    model.compile({ loss: 'meanSquaredError', optimizer: 'adam' });

    return model;
}

// Function to save the model
async function saveModel(model, modelPath) {
    const saveResult = await model.save(`file://${modelPath}`);
    console.log('Model saved:', saveResult);
}

// Function to load the model
async function loadModel(modelPath) {
    const loadedModel = await tf.loadLayersModel(`file://${modelPath}`);
    console.log('Model loaded successfully');
    return loadedModel;
}

// Function to make predictions using the loaded model
function predict(model, inputData, scaler) {
    // Normalize the input data using the scaler
    const normalizedInput = scaler.apply(tf.tensor2d([inputData]));

    // Reshape the input tensor to be 3D [batch_size, timesteps, features]
    const inputTensor = normalizedInput.reshape([1, normalizedInput.shape[0], 1]);

    // Make predictions using the loaded model
    const prediction = model.predict(inputTensor);
    const predictionData = prediction.arraySync();

    // Invert predictions using the scaler
    const invertedPrediction = scaler.inverseTransform(tf.tensor2d(predictionData, [predictionData.length, 1])).arraySync();

    console.log('Inverted Prediction:', invertedPrediction);
}

// Train the model
async function trainModel(model, trainX, trainY, epochs = 10, batchSize = 1) {
    await model.fit(trainX, trainY, {
        epochs,
        batchSize,
        verbose: 1,
    });
}

// Function to make predictions and evaluate accuracy with graph
function evaluateModel(model, testX, testY, scaler, lookBack, dataset, trainX, trainY) {
    const testPredict = model.predict(testX).arraySync();
    const trainPredict = model.predict(trainX).arraySync()
    const testYArray = testY.arraySync();
    const trainYArray = trainY.arraySync();

    // Invert predictions
    const invertedTestPredict = scaler
        .inverseTransform(tf.tensor2d(testPredict, [testPredict.length, 1]))
        .arraySync();

    // Invert true values
    const invertedTestY = scaler
        .inverseTransform(tf.tensor2d(testYArray, [testYArray.length, 1]))
        .arraySync();

    // Invert predictions
    const invertedTrainPredict = scaler
        .inverseTransform(tf.tensor2d(trainPredict, [trainPredict.length, 1]))
        .arraySync();

    // Invert true values
    const invertedTrainY = scaler
        .inverseTransform(tf.tensor2d(trainYArray, [trainYArray.length, 1]))
        .arraySync();

    // Calculate custom accuracy
    const accuracy = customAccuracy(invertedTestY, invertedTestPredict);
    console.log(`Test Accuracy: ${accuracy.toFixed(2)}`);

    // console.log(invertedTestPredict, invertedTestY, invertedTrainPredict, invertedTrainY)

    return { invertedTestPredict, invertedTestY };
}

// Custom accuracy function
function customAccuracy(yTrue, yPred) {
    const percentageDiff = yTrue.map((trueVal, index) => {
        return Math.abs((trueVal - yPred[index]) / trueVal) * 100;
    });

    const accuracy =
        percentageDiff.filter(diff => diff <= 15).length / percentageDiff.length;

    return accuracy;
}

// Function to create datasets
function createDatasets(dataset, lookBack) {
    const dataX = [];
    const dataY = [];

    for (let i = 0; i < dataset.length - lookBack - 1; i++) {
        const a = dataset.slice(i, i + lookBack);
        const b = dataset[i + lookBack];

        dataX.push(a);
        dataY.push(b);
    }

    const splitIndex = Math.floor(dataset.length * 0.8);

    const trainX = tf.tensor3d(dataX.slice(0, splitIndex), [splitIndex, lookBack, 1]);
    const trainY = tf.tensor2d(dataY.slice(0, splitIndex), [splitIndex, 1]);

    const testX = tf.tensor3d(dataX.slice(splitIndex), [dataset.length - splitIndex - lookBack - 1, lookBack, 1]);
    const testY = tf.tensor2d(dataY.slice(splitIndex), [dataset.length - splitIndex - lookBack - 1, 1]);

    return [trainX, trainY, testX, testY];
}

// Save scaler as JSON
function saveScaler(scaler, scalerPath) {
    const scalerData = scaler.getScalerData();
    const scalerJson = JSON.stringify(scalerData);

    fs.writeFileSync(scalerPath, scalerJson);
    console.log('Scaler saved:', scalerPath);
}

// Load scaler from JSON
function loadScaler(scalerPath) {
    const scalerJson = fs.readFileSync(scalerPath, 'utf-8');
    const scalerData = JSON.parse(scalerJson);

    // Create a CustomScaler instance based on the loaded data
    const scaler = new CustomScaler(tf.scalar(scalerData.min), tf.scalar(scalerData.max));

    console.log('Scaler loaded successfully');
    return scaler;
}

// Main execution
async function main() {
    const { dataset, scaler } = await loadDataset();

    const lookBack = 1;
    const [trainX, trainY, testX, testY] = createDatasets(dataset.arraySync(), lookBack);

    const model = createModel([lookBack, 1]);
    await trainModel(model, trainX, trainY);

    // Save the trained model
    const modelPath = "models/model"
    await saveModel(model, modelPath);

    // Save scaler
    const scalerPath = 'models/model/scaler.json';
    saveScaler(scaler, scalerPath);

    evaluateModel(model, testX, testY, scaler, lookBack, dataset, trainX, trainY);
}

// 
async function predictMain() {
    const modelPath = "./models/model/model.json"
    const scalerPath = "./models/model/scaler.json"

    // Load the saved model
    const loadedModel = await loadModel(modelPath);
    const loadedScaler = loadScaler(scalerPath);

    // Example input data for prediction
    const inputData = [355];

    // Make predictions using the loaded model
    predict(loadedModel, inputData, loadedScaler);
}

main();
// predictMain()

File: sales.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LSTM
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error

# # fix random seed for reproducibility
tf.random.set_seed(7)

# load the dataset
dataframe = pd.read_csv('airline-passengers.csv', usecols=[1], engine='python')
dataset = dataframe.values
dataset = dataset.astype('float32')

# normalize the dataset
scaler = MinMaxScaler(feature_range=(0, 1))
dataset = scaler.fit_transform(dataset)

# split into train and test sets
train_size = int(len(dataset) * 0.67)
test_size = len(dataset) - train_size
train, test = dataset[0:train_size,:], dataset[train_size:len(dataset),:]
print(len(train), len(test))

# convert an array of values into a dataset matrix
def create_dataset(dataset, look_back=1):
	dataX, dataY = [], []
	for i in range(len(dataset)-look_back-1):
		a = dataset[i:(i+look_back), 0]
		dataX.append(a)
		dataY.append(dataset[i + look_back, 0])
	return np.array(dataX), np.array(dataY)

# reshape into X=t and Y=t+1
look_back = 1
trainX, trainY = create_dataset(train, look_back)
testX, testY = create_dataset(test, look_back)

print(trainX,trainY)


# reshape input to be [samples, time steps, features]
trainX = np.reshape(trainX, (trainX.shape[0], 1, trainX.shape[1]))
testX = np.reshape(testX, (testX.shape[0], 1, testX.shape[1]))


# create and fit the LSTM network
model = Sequential()
model.add(LSTM(4, input_shape=(1, look_back)))
model.add(Dense(1))
model.compile(loss='mean_squared_error', optimizer='adam')
model.fit(trainX, trainY, epochs=100, batch_size=1, verbose=0)


# calculate accuracy
accuracy = model.evaluate(testX, testY, verbose=0)
print('Test Accuracy: %.2f' % (accuracy))


# make predictions
trainPredict = model.predict(trainX)
testPredict = model.predict(testX)
# invert predictions
trainPredict = scaler.inverse_transform(trainPredict)
trainY = scaler.inverse_transform([trainY])
testPredict = scaler.inverse_transform(testPredict)
testY = scaler.inverse_transform([testY])
# calculate root mean squared error
trainScore = np.sqrt(mean_squared_error(trainY[0], trainPredict[:,0]))
print('Train Score: %.2f RMSE' % (trainScore))
testScore = np.sqrt(mean_squared_error(testY[0], testPredict[:,0]))
print('Test Score: %.2f RMSE' % (testScore))


def custom_accuracy(y_true, y_pred):
    # calculate percentage difference between predicted and original data
	percentage_diff = np.abs((y_true - y_pred) / y_true) * 100
    
    # calculate accuracy as the inverse proportion of samples with percentage difference <= 10%
	accuracy = np.sum(percentage_diff <= 10) / len(y_true)
    
	return accuracy

# calculate custom accuracy
train_accuracy = custom_accuracy(trainY[0], trainPredict[:,0])
test_accuracy = custom_accuracy(testY[0], testPredict[:,0])

print('Train Accuracy: %.2f' % (train_accuracy))
print('Test Accuracy: %.2f' % (test_accuracy))


# shift train predictions for plotting
trainPredictPlot = np.empty_like(dataset)
trainPredictPlot[:, :] = np.nan
trainPredictPlot[look_back:len(trainPredict)+look_back, :] = trainPredict
# shift test predictions for plotting
testPredictPlot = np.empty_like(dataset)
testPredictPlot[:, :] = np.nan
testPredictPlot[len(trainPredict)+(look_back*2)+1:len(dataset)-1, :] = testPredict
# plot baseline and predictions
plt.plot(scaler.inverse_transform(dataset), label='Baseline')
plt.plot(trainPredictPlot, label='Train Predictions')
plt.plot(testPredictPlot, label='Test Predictions')
plt.legend()
plt.show()


# Function to predict with the loaded scaler
def predict_with_scaler(scaler, input_data):
    # Normalize input data with the loaded scaler
    normalized_data = scaler.transform(input_data)

    # Reshape normalized data
    normalized_data = np.reshape(normalized_data, (normalized_data.shape[0], 1, normalized_data.shape[1]))

    # Make predictions with the original model
    predictions = model.predict(normalized_data)

    # Invert predictions with the loaded scaler
    inverted_predictions = scaler.inverse_transform(predictions)

    return inverted_predictions

new_data = np.array([[231]])  # Replace with your own data
predicted_values = predict_with_scaler(scaler, new_data)
print("Predicted Values:", predicted_values)